### PR TITLE
Add Contributions flow - Initial

### PR DESCRIPTION
**Pull request by dparikh**

### CODE REVIEW

File: src/runtime/contributions-flow.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+    /** @private @const {!web-activities/activity-ports.ActivityPorts} */
+    this.activityPorts_ = deps.activities();
+
+    /** @private @const {!../components/dialog-manager.DialogManager} */
+    this.dialogManager_ = deps.dialogManager();
+
+    let isClosable = options && options.isClosable;
+    if (isClosable == undefined) {
+      isClosable = true;  // Default is to show the Close button.
```
Comment (sohanirao) [inline src/runtime/contributions-flow.js]: why not just do isCloseble = (options && options.isCloseable) || true;

**Comment by dparikh** [comment]
Added doc.

File: docs/contributions-flow.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+# SwG Contributions Flow
+
+This flow allows the publication site to display a flow to contribute to
+the publisher. See [Subscriptions APIs](./core-apis.md).
+
+*Important!* Please ensure you set up the `setOnSubscribeResponse` on any page where you accept purchases, not just before you call `subscribe` or `showContributions`. SwG client ensures it can recover contributions even when browsers unload pages. See [Subscribe flow](./subscribe-flow.md) for more details.
+
+The contribution flow will first present a set of amounts user can contribute to.
+A user will get a choice to either select one of the offers, or try request login to claim an existing contribution.
```
Comment (sohanirao) [inline docs/contributions-flow.md]: can we reword it to ... 
...offers, or login to claim an existing contributions.

File: docs/contributions-flow.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+
+## Contribution options
+
+The contributions API (`showContributions`) accept a list of SKUs to be displayed.
+
+For instance:
+
+```
+subscriptions.showContributions({skus: ['sku1', 'sku2']});
```
Comment (sohanirao) [inline docs/contributions-flow.md]: is there any way to combine this description in Line 27-30:
Like say
To display contributions, use the showContributions() API. This API doesn't need any arguments but can be called optionally with a list of SKUs to be displayed.
opt_args = {skus: ['sku1', 'sku2']}; // or undefined
subscriptions.showContributions(opt_args);

File: docs/contributions-flow.md (hunk 2 above)
Comment (dparikh) [reply docs/contributions-flow.md]: I think, "request login" makes little more clear.

File: docs/contributions-flow.md (hunk 3 above)
Comment (dparikh) [reply docs/contributions-flow.md]: This is consistent with the Offers flow:

https://github.com/subscriptions-project/swg-js/blob/master/docs/offers-flow.md

File: docs/contributions-flow.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+distributed under the License is distributed on an "AS-IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+# SwG Contributions Flow
+
+This flow allows the publication site to display a flow to contribute to
+the publisher. See [Subscriptions APIs](./core-apis.md).
```
Comment (chenshay) [inline docs/contributions-flow.md]: Change to: "This flow allows the publication site to display a flow where users can contribute money to the publisher"

File: docs/contributions-flow.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+*Important!* Please ensure you set up the `setOnSubscribeResponse` on any page where you accept purchases, not just before you call `subscribe` or `showContributions`. SwG client ensures it can recover contributions even when browsers unload pages. See [Subscribe flow](./subscribe-flow.md) for more details.
```
Comment (chenshay) [inline docs/contributions-flow.md]: This is confusing. I personally don't understand exactly what it's referring to and expect that an outside developer would have a hard time as well.

Comment (chenshay) [reply docs/contributions-flow.md]: This probably belongs much lower in the doc

File: docs/contributions-flow.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+To handle the login request:
+
+```
+subscriptions.setOnLoginRequest(function() {
+  // Handle login request.
+});
+```
+
+If a user elects for a presented offer, SwG will run the [Subscribe flow](./subscribe-flow.md).
```
Comment (chenshay) [inline docs/contributions-flow.md]: SwG is an acronym that isn't recognized by all developers. I'd refer directly to the script file.

File: docs/contributions-flow.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+
+## Contribution options
+
+The contributions API (`showContributions`) accept a list of SKUs to be displayed.
```
Comment (chenshay) [inline docs/contributions-flow.md]: accept -> accepts

File: docs/contributions-flow.md (hunk 4 above)
Comment (dparikh) [reply docs/contributions-flow.md]: Done

File: docs/contributions-flow.md (hunk 5 above)
Comment (dparikh) [reply docs/contributions-flow.md]: I would leave this as is for now to be consistent with Offers flow.

We could discuss this how we could improve the docs overall.

File: docs/contributions-flow.md (hunk 7 above)
Comment (dparikh) [reply docs/contributions-flow.md]: Done

File: docs/contributions-flow.md (hunk 6 above)
Comment (dparikh) [reply docs/contributions-flow.md]: Same for previous comment.
We could revisit our docs for SwG to make improvements.

File: src/api/subscriptions.js [hunk 8]
```diff
@@ -115,6 +115,12 @@ export class Subscriptions {
    */
   subscribe(skuOrSubscriptionRequest) {}
 
+  /**
+   * Starts the Contributions flow.
+   * @param {!OffersRequest=} opt_options
+   */
+  showContributions(opt_options) {}
```
Comment (dvoytenko) [inline src/api/subscriptions.js]: Seems like the name is not very clear. We are not showing contributions here themselves, but rather contributions options?

File: src/api/subscriptions.js [hunk 9]
```diff
@@ -115,6 +115,12 @@ export class Subscriptions {
    */
   subscribe(skuOrSubscriptionRequest) {}
 
+  /**
+   * Starts the Contributions flow.
```
Comment (dvoytenko) [inline src/api/subscriptions.js]: I'm not entirely clear we are ready to define top-level API here. Can we defer this until we have the flow a bit more flashed out? Alternatively, we should put it under the experiment. Otherwise, I'd rather not have the new unprotected API appearing in the PROD API.

File: src/api/subscriptions.js (hunk 8 above)
Comment (dparikh) [reply src/api/subscriptions.js]: Removed for now.

File: src/api/subscriptions.js (hunk 9 above)
Comment (dparikh) [reply src/api/subscriptions.js]: Removed for now from the API.

File: docs/contributions-flow.md (hunk 2 above)
Comment (dparikh) [reply docs/contributions-flow.md]: This is same as with OffersView whereas if user is not logged-in need to login-in to fetch the entitlements.
Again, the wordings are same as with OffersView.

File: docs/contributions-flow.md (hunk 3 above)
Comment (dparikh) [reply docs/contributions-flow.md]: Not sure, if I understood your previous comment and this new one since, it appears to be contradicting.

Happy to chat offline if you have other ideas.

File: docs/contributions-flow.md (hunk 5 above)
Comment (chenshay) [reply docs/contributions-flow.md]: I would move this to the bottom and create an issue to improve the Offers flow to do the same

Comment (dparikh) [reply docs/contributions-flow.md]: Done.
